### PR TITLE
feat: smart routing via IPVS

### LINKED ISSUE #91: NAT bypassing
After struggling for a couple of days I've came to the idea how we can allow miners without public IPs to be able to work.

I've made a prototype and it works! Multiple PRs incoming ...

Le plan:

- [x] Detect NAT type on the miner.
  - If there is no NAT type or it's weak like symmetric UDP firewall - that's fine. Allow them to publish themselves on their white IPs.
  - If there is NAT or some other firewall which restricts public access - then we're in trouble.
- [x] Provide this information to the Hub.
- [x] Hub can organise virtual private network with such miners, allowing requests to be reachable at least from the Hub's endpoint. Either directly or using NAT piercing strategy with further ipip tunnelling.
- [x] For miners with public IPs there will be direct access with containers. That information is provided using **start** and **info** requests.
- [x] For miners that can't be reachable from the WAN we can provide dynamic routing using IPVS configured as NAT. Then provide our own public IP and remapped port for virtual services as an endpoint(s) to the task.

Sudden profits:
- Smart and configurable (rr, lwc, wrr, etc.) load balancing between similar containers from the same task. For now we can have only 1 container per task, however it is possible to perform auto-scaling.
- 1:1 port mapping is also becomes possible, especially for applications that can't be configured.
- Both TCP/UDP routing.

Pros:
- Fast, as it is L3 forwarding and is done in the kernel.
- Fully optional. Some kind as extension for miners that can't obtain public IP for some reasons.

Cons:
- Miner should set the Hub as a default gateway (or precisely for a given restricted port range using iptables), otherwise packets will be dropped by the kernel as unexpected, even if the client is reachable directly from miners.
- Routed traffic must be additionally accounted (however it's possible to). Not sure it's cons :).
- Increased timings (not much), as it's an additional hop between clients and miners.
- Hubs are restricted to be run on linux machine, because of netlink and lvs.

### PR DESCRIPTION
**Pull request by 3Hren**
This PR introduces traffic routing for Miners under the firewall using IPVS.

TBD:
- [x] Route Docker network properly.
- [x] Dismiss virtual services on task stop.
- [x] Dismiss virtual services on shutdown.
- [x] Return properly remapped endpoints on task start reply.

Closes #91.

### CODE REVIEW

File: insonmnia/gateway/pool.go [hunk 1]
```diff
@@ -0,0 +1,60 @@
+package gateway
+
+import (
+	"github.com/pkg/errors"
```
Comment (noxiouz) [inline insonmnia/gateway/pool.go]: just errors, I suppose

File: insonmnia/miner/marshal.go [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+	"github.com/ccding/go-stun/stun"
+
+	pb "github.com/sonm-io/core/proto"
+)
+
+func marshalNATType(nat stun.NATType) pb.NATType {
+	switch nat {
+	case stun.NATNone:
+		return pb.NATType_NONE
+	case stun.NATError:
```
Comment (noxiouz) [inline insonmnia/miner/marshal.go]: case stun.NATError, stun.NATUnknown:
....

File: insonmnia/gateway/pool.go [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+
+	return port.(uint16), nil
+}
+
+func (p *PortPool) Retain(ID string) error {
+	p.mu.Lock()
+	defer p.mu.Unlock()
+
+	if port, exists := p.used[ID]; exists {
+		p.queue.Enqueue(port)
```
Comment (noxiouz) [inline insonmnia/gateway/pool.go]: !exists return error

p..Enqueue
delete
...

Comment (3Hren) [inline insonmnia/gateway/pool.go]: We need a port to enqueue and to check in a single map touching.

**Comment by 3Hren** [comment]
That's all. Discuss!

File: insonmnia/gateway/gateway_linux.go [hunk 4]
```diff
@@ -0,0 +1,243 @@
...
+		zap.String("service", vsID),
+		zap.String("backend", rsID),
+		zap.String("host", host),
+		zap.Uint16("port", port),
+	)
+
+	if err := g.ipvs.DelDestPort(host, port, backendHost, backendPort, protocol); err != nil {
+		log.G(g.ctx).Error("failed to remove backend",
+			zap.String("service", vsID),
+			zap.String("backend", rsID),
```
Comment (noxiouz) [inline insonmnia/gateway/gateway_linux.go]: zap.Error(err)

**Pull request merged**